### PR TITLE
skip hardlinking inverted index files in mutation

### LINKED ISSUE #47393: inverted index bug on segment id
**Describe what's wrong**
As in title.

**Does it reproduce on recent release?**
Yes.

**How to reproduce**

```sql
set allow_experimental_inverted_index=1;
CREATE TABLE t
(
    `timestamp` UInt64,
    `s` String,
    INDEX idx lower(s) TYPE inverted(3) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY toDate(timestamp) -- partitioning may be necessary
ORDER BY toDate(timestamp)
SETTINGS min_rows_for_wide_part = 1, min_bytes_for_wide_part = 1; -- these settings are necessary

INSERT INTO t (s) SELECT * FROM generateRandom('s String') LIMIT 100;

-- do update column
ALTER TABLE t UPDATE flag=1 WHERE 1;
-- wait for mutation to finish on at least 1 part
SELECT sleepEachRow(1) FROM t LIMIT 3;

SELECT * FROM t WHERE lower(s) LIKE '%iamok%';
```
see [ClickFiddle](https://fiddle.clickhouse.com/8eb0e0ed-8ec2-4326-8880-3ecb0bec1747)

result:
```
[chi-datalake-ck-cluster-2-0-0] 2023.03.09 16:34:23.859618 [ 388 ] {f78d765e-4cec-4cfd-a49c-6379667f915b} <Error> executeQuery: Code: 49. DB::Exception: Invalid segment id 1. (LOGICAL_ERROR) (version 23.2.3.17 (official build)) (from 127.0.0.1:43318) (in query: SELECT * FROM t WHERE lower(s) LIKE '%iamnotok%'), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0xe0c67d5 in /usr/bin/clickhouse
1. ? @ 0xee7502f in /usr/bin/clickhouse
2. DB::GinIndexStoreDeserializer::readSegmentDictionary(unsigned int) @ 0x141f106a in /usr/bin/clickhouse
3. DB::GinIndexStoreFactory::get(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::IDataPartStorage const>) @ 0x141f2e36 in /usr/bin/clickhouse
4. DB::MergeTreeDataSelectExecutor::filterMarksUsingIndex(std::__1::shared_ptr<DB::IMergeTreeIndex const>, std::__1::shared_ptr<DB::IMergeTreeIndexCondition>, std::__1::shared_ptr<DB::IMergeTreeDataPart const>, DB::MarkRanges const&, DB::Settings const&, DB::MergeTreeReaderSettings const&, unsigned long&, unsigned long&, DB::MarkCache*, DB::UncompressedCache*, Poco::Logger*) @ 0x14330bcc in /usr/bin/clickhouse
5. ? @ 0x1432d83d in /usr/bin/clickhouse
6. DB::MergeTreeDataSelectExecutor::filterPartsByPrimaryKeyAndSkipIndexes(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>&&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const>, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const> const&, DB::KeyCondition const&, DB::MergeTreeReaderSettings const&, Poco::Logger*, unsigned long, std::__1::vector<DB::ReadFromMergeTree::IndexStat, std::__1::allocator<DB::ReadFromMergeTree::IndexStat>>&, bool) @ 0x1432b1bb in /usr/bin/clickhouse
7. DB::ReadFromMergeTree::selectRangesToReadImpl(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const>, unsigned long, std::__1::shared_ptr<std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, long>>>>, DB::MergeTreeData const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, bool, Poco::Logger*) @ 0x14aea48e in /usr/bin/clickhouse
8. DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>, std::__1::shared_ptr<DB::PrewhereInfo> const&, DB::ActionDAGNodes const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const>, unsigned long, std::__1::shared_ptr<std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const, long>>>>, DB::MergeTreeData const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>>> const&, bool, Poco::Logger*) @ 0x14ae875f in /usr/bin/clickhouse
9. DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const>>>) const @ 0x14ae7366 in /usr/bin/clickhouse
10. DB::ReadFromMergeTree::getAnalysisResult() const @ 0x14aed196 in /usr/bin/clickhouse
11. DB::ReadFromMergeTree::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&) @ 0x14aef2a9 in /usr/bin/clickhouse
12. DB::ISourceStep::updatePipeline(std::__1::vector<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>, std::__1::allocator<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder>>>>, DB::BuildQueryPipelineSettings const&) @ 0x14ab4814 in /usr/bin/clickhouse
13. DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&) @ 0x14ace149 in /usr/bin/clickhouse
14. DB::InterpreterSelectWithUnionQuery::execute() @ 0x1365c72d in /usr/bin/clickhouse
15. ? @ 0x1397b559 in /usr/bin/clickhouse
16. DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>> const&, std::__1::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x1397862d in /usr/bin/clickhouse
17. DB::TCPHandler::runImpl() @ 0x146ff144 in /usr/bin/clickhouse
18. DB::TCPHandler::run() @ 0x14713979 in /usr/bin/clickhouse
19. Poco::Net::TCPServerConnection::start() @ 0x17614454 in /usr/bin/clickhouse
20. Poco::Net::TCPServerDispatcher::run() @ 0x1761567b in /usr/bin/clickhouse
21. Poco::PooledThread::run() @ 0x1779ca07 in /usr/bin/clickhouse
22. Poco::ThreadImpl::runnableEntry(void*) @ 0x1779a43d in /usr/bin/clickhouse
23. ? @ 0x7f8809b7d609 in ?
24. clone @ 0x7f8809aa2133 in ?


0 rows in set. Elapsed: 0.004 sec.

Received exception from server (version 23.2.3):
Code: 49. DB::Exception: Received from localhost:9000. DB::Exception: Invalid segment id 1. (LOGICAL_ERROR)
```

### PR DESCRIPTION
**Pull request by ardenwick**
In mutation task, files that are not affected are hardlinked into new part, for wide part.

as I understand, If the inverted index be recalculated, their files should not be linked to the new part.

Close #47393

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in official stable or prestable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
fix invalid segment id bug after mutation

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by cangyin** [comment]
some tests failed, but looks irrelavant. what should I do next ?

File: tests/queries/0_stateless/02346_inverted_index_mutation.sql [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ORDER BY tuple()
+SETTINGS min_rows_for_wide_part = 1, min_bytes_for_wide_part = 1;
+
+INSERT INTO t (s) VALUES ('I am inverted');
+
+-- do update column
+ALTER TABLE t UPDATE s='I am not inverted' WHERE 1;
+
+-- wait for mutation to finish on at least 1 part
+SELECT sleepEachRow(1) FROM t LIMIT 1;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/02346_inverted_index_mutation.sql:20]: It does not look reliable.

Comment (cangyin) [reply tests/queries/0_stateless/02346_inverted_index_mutation.sql:20]: is there some settings or statements for this ?

Comment (cangyin) [reply tests/queries/0_stateless/02346_inverted_index_mutation.sql]: I found `mutations_sync` to be a better way

**Comment by cangyin** [comment]
finally, all checks turned green in actions page https://github.com/ClickHouse/ClickHouse/actions/runs/4454242025/jobs/7825239544

but a 'Stress test (debug)' in this PR page is failing, is that ok ?

**Comment by alexey-milovidov** [comment]
The reports of our CI have poor usability for the following reasons:

1. The checks are run with GitHub Actions on "spot" (interruptible) instances in AWS. When the instance where the test is run is interrupted, it will be shown as "PullRequestCI" failure, but the check will be restarted. A few hours after the commit, you will see annoying and misleading failed checks from the "PullRequestCI", that are restarted and fixed automatically.

2. Every check is presented in the list on GitHub twice: as a GitHub Actions run (like "PullRequestCI ...", usually located at top) and as a generated report to look (like "ClickHouse build check"). The generated report presents the test results, while the PullRequestCI presents the GitHub actions logs.

3. There are dependencies between checks - for example, if Fast Test fails, the more heavy Build Check will not run, but in this case, a report like "ClickHouse build check — 0/8 artifact groups are OK (0 of 8 builds are OK)" will be published, but it does not give any information, and it's even misleading (you can click there by mistake and waste a minute of your time).

These reasons make me sorry for the fact that the new external contributors are overwhelmed by the reports that are looking randomly and don't know where to look for the proper results.
Note: it is still perfectly usable if you know where to click and get used to it; and in fact, our CI is (as far as I know) the most powerful among open-source with publicly available results.

To overcome the poor usability of GitHub actions and to fix this problem, @Felixoid is going to try to introduce a summary of the checks in a comment with a report, that will be updated automatically.

**Comment by cangyin** [comment]
seems ClickHouse may need a better way to handle additional files for a index.

**Comment by cangyin** [comment]
@alexey-milovidov  @rschu1ze Thanks for taking the time to review my PR. I was wondering if there are any outstanding issues that need to be addressed before the PR can be merged, and when I can expect a decision on whether or not it will be merged ?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by robot-clickhouse-ci-2** [comment]
*This is an automated comment for commit 885e07cd0b748c4a0ffbb6f2bcdf86e521bf00ef with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/47663/885e07cd0b748c4a0ffbb6f2bcdf86e521bf00ef/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td><a href="https://github.com/ClickHouse/ClickHouse/actions/runs/6450639779">✅ success</a></td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/47663/885e07cd0b748c4a0ffbb6f2bcdf86e521bf00ef/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Pull request merged**